app/src/backend/ActivityService.ts

```diff
@@ -21,9 +21,241 @@ import type { Transaction } from "sequelize";
 
 type GasValueInput = Omit<GasValueCreationAttributes, "id"> & {
   emissionsFactor?: Omit<EmissionsFactorAttributes, "id">;
+  id?: string;
 };
 
 export default class ActivityService {
+  private static async updateDataSource({
+    activityValue,
+    dataSourceParams,
+    transaction,
+  }: {
+    activityValue: ActivityValue;
+    dataSourceParams: Omit<DataSourceAttributes, "datasourceId"> | undefined;
+    transaction: Transaction;
+  }): Promise<string> {
+    if (activityValue.datasourceId) {
+      const dataSource = await db.models.DataSource.findOne({
+        where: { datasourceId: activityValue.datasourceId },
+        transaction,
+      });
+      if (!dataSource) {
+        throw new createHttpError.NotFound(
+          "Data source for ActivityValue not found",
+        );
+      }
+      Object.assign(dataSource, dataSourceParams);
+      await dataSource?.save({ transaction });
```

Comment:
Sequelize has a `.update()` method you can use here to simplify the logic a bit.
You can just pass the dataSourceParams and the existing fields will stay the same.
There is also `.upsert()` if you're not sure if it exists yet, but I think here you can assume it does.